File: tests/csharp/src/T012_Specification.cs
namespace T012
{
	using GMSEC.API5;
	using GMSEC.API5.Testing.Common;
	using System;
    using System.Collections.Generic;
    using System.Text;


	class T012_Specification : TestCase
	{
		static void Main(string[] args)
		{
			T012_Specification test = new T012_Specification();
			test.SetDescription("Test Specification");
			test.Run(args);
		}


		public override void Run()
		{
			Test_Constructor_1();
			Test_Constructor_2();
			Test_CopyConstructor();
			Test_GetSchemaIDIterator();
			Test_GetMessageSpecifications();
			Test_GetHeaderFieldNames();
		}


		private void Test_Constructor_1()
		{
			Log.Info("Test constructor (no args)");

			Specification spec = new Specification();
			Check("Unexpected specification version", spec.GetVersion() == Gmsec.GMSEC_MSG_SPEC_CURRENT);
			Check("Unexpected specification schema level", spec.GetSchemaLevel() == Specification.SchemaLevel.LEVEL_0);
		}


		private void Test_Constructor_2()
		{
			Log.Info("Test constructor w/ Config");

			//o Nominal tests
			{
				Specification spec = new Specification( new Config() );
				Check("Unexpected specification version", spec.GetVersion() == Gmsec.GMSEC_MSG_SPEC_CURRENT);
				Check("Unexpected specification schema level", spec.GetSchemaLevel() == Specification.SchemaLevel.LEVEL_0);
			}

			{
				Specification spec0 = new Specification( new Config("gmsec-specification-version=201900 gmsec-schema-level=0", DataType.KEY_VALUE_DATA) );
				Specification spec1 = new Specification( new Config("gmsec-specification-version=201900 gmsec-schema-level=1", DataType.KEY_VALUE_DATA) );
				Specification spec2 = new Specification( new Config("gmsec-specification-version=201900 gmsec-schema-level=2", DataType.KEY_VALUE_DATA) );

				Check("Unexpected specification version", Gmsec.GMSEC_MSG_SPEC_2019_00 == spec0.GetVersion());
				Check("Unexpected specification schema level", Specification.SchemaLevel.LEVEL_0 == spec0.GetSchemaLevel());

				Check("Unexpected specification version", Gmsec.GMSEC_MSG_SPEC_2019_00 == spec1.GetVersion());
				Check("Unexpected specification schema level", Specification.SchemaLevel.LEVEL_1 == spec1.GetSchemaLevel());

				Check("Unexpected specification version", Gmsec.GMSEC_MSG_SPEC_2019_00 == spec2.GetVersion());
				Check("Unexpected specification schema level", Specification.SchemaLevel.LEVEL_2 == spec2.GetSchemaLevel());
			}

			{
				Specification spec = new Specification( new Config("name=value", DataType.KEY_VALUE_DATA) );
				Check("Unexpected specification version", spec.GetVersion() == Gmsec.GMSEC_MSG_SPEC_CURRENT);
				Check("Unexpected specification schema level", spec.GetSchemaLevel() == Specification.SchemaLevel.LEVEL_0);
			}

			//o Off-nominal tests
			try {
				new Specification( new Config("gmsec-specification-version=201300", DataType.KEY_VALUE_DATA) );
				Check("An exception was expected", false);
			}
			catch (GmsecException e) {
				Check(e.ToString(), e.ToString().Contains("Could not list files in template directory"));
			}

			try {
				new Specification( new Config("gmsec-specification-version=ABC", DataType.KEY_VALUE_DATA) );
				Check("An exception was expected", false);
			}
			catch (GmsecException e) {
				Check(e.ToString(), e.ToString().Contains("\"GMSEC-SPECIFICATION-VERSION\" contains invalid value"));
			}

			try {
				new Specification( new Config("gmsec-schema-level=3", DataType.KEY_VALUE_DATA) );
				Check("An exception was expected", false);
			}
			catch (GmsecException e) {
				Check(e.ToString(), e.ToString().Contains("DIRECTORY is missing definition for LEVEL-3 HEADER"));
			}

			try {
				new Specification( new Config("gmsec-schema-level=10", DataType.KEY_VALUE_DATA) );
				Check("An exception was expected", false);
			}
			catch (GmsecException e) {
				Check(e.ToString(), e.ToString().Contains("\"GMSEC-SCHEMA-LEVEL\" contains out-of-range integer"));
			}

			try {
				new Specification( new Config("gmsec-schema-level=ABC", DataType.KEY_VALUE_DATA) );
				Check("An exception was expected", false);
			}
			catch (GmsecException e) {
				Check(e.ToString(), e.ToString().Contains("\"GMSEC-SCHEMA-LEVEL\" contains invalid value"));
			}

			try {
				new Specification( new Config("gmsec-schema-path=here", DataType.KEY_VALUE_DATA) );
				Check("An exception was expected", false);
			}
			catch (GmsecException e) {
				Check(e.ToString(), e.ToString().Contains("Could not list files in template directory"));
			}
		}


		private void Test_CopyConstructor()
		{
			Log.Info("Test copy-constructor");

			Specification spec1 = new Specification();
			Specification spec2 = new Specification(spec1);

			Check("Unexpected specification version", spec1.GetVersion() == spec2.GetVersion());
			Check("Unexpected specification schema level", spec1.GetSchemaLevel() == spec2.GetSchemaLevel());
			Check("Unexpected message specifications", spec1.GetMessageSpecifications().Count == spec2.GetMessageSpecifications().Count);
		}


		private void Test_GetSchemaIDIterator()
		{
			Log.Info("Test GetSchemaIDIterator()");

			Specification spec = new Specification(); 

			SchemaIDIterator iter = spec.GetSchemaIDIterator();

			Check("Expected availability of schema IDs", iter.HasNext());
		}


		private void Test_GetMessageSpecifications()
		{
			Log.Info("Test GetMessageSpecifications()");

			Specification spec = new Specification();

			MessageSpecificationList msgSpecs = spec.GetMessageSpecifications();

			Check("Expected to have message specifications", msgSpecs.Count > 0);
		}


		private void Test_GetHeaderFieldNames()
		{
			Log.Info("Test GetHeaderFieldNames()");

			Specification spec = new Specification();

			HeaderList headers = spec.GetHeaderFieldNames();
			Check("Expected to have headers", headers.Count == 26);

			HeaderList headers2 = spec.GetHeaderFieldNames("");
			Check("Expected to have headers", headers2.Count == 26);

			HeaderList defaultHeaders = spec.GetHeaderFieldNames("DEFAULT");
			Check("Expected to have headers", defaultHeaders.Count == 26);

			HeaderList C2MSHeaders = spec.GetHeaderFieldNames("C2MS");
			Check("Expected to have headers", C2MSHeaders.Count == 26);
		}
	}
}

File: tests/java/src/T012_Specification.java
import gov.nasa.gsfc.gmsec.api5.Config;
import gov.nasa.gsfc.gmsec.api5.Gmsec;
import gov.nasa.gsfc.gmsec.api5.GmsecException;
import gov.nasa.gsfc.gmsec.api5.MessageSpecification;
import gov.nasa.gsfc.gmsec.api5.SchemaIDIterator;
import gov.nasa.gsfc.gmsec.api5.Specification;

import gov.nasa.gsfc.gmsec.api5.util.Log;

import java.util.Collection;


public class T012_Specification extends TestCase
{
	public static void main(String[] args)
	{
		T012_Specification test = new T012_Specification();
		test.setDescription("Test Specification");
		test.run(args);
	}


    public void testCase()
		throws Exception
	{
		try
		{
			test_constructor_1();
			test_constructor_2();
			test_copy_constructor();
			test_get_schema_id_iterator();
			test_get_message_specifications();
			test_get_header_field_names();
		}
		catch (Exception e)
		{
			require(e.getMessage(), false);
		}
	}


	private void test_constructor_1()
		throws Exception
	{
		Log.info("Test constructor (no args)");

		Specification spec = new Specification();
		check("Unexpected specification version", spec.getVersion() == Gmsec.MSG_SPEC_CURRENT);
		check("Unexpected specification schema level", spec.getSchemaLevel() == Specification.SchemaLevel.LEVEL_0);
	}


	private void test_constructor_2()
		throws Exception
	{
		Log.info("Test constructor w/ Config");

		//o Nominal tests
		{
			Specification spec = new Specification( new Config() );
			check("Unexpected specification version", spec.getVersion() == Gmsec.MSG_SPEC_CURRENT);
			check("Unexpected specification schema level", spec.getSchemaLevel() == Specification.SchemaLevel.LEVEL_0);
		}

		{
			Specification spec0 = new Specification( new Config("gmsec-specification-version=201900 gmsec-schema-level=0", Gmsec.DataType.KEY_VALUE_DATA) );
			Specification spec1 = new Specification( new Config("gmsec-specification-version=201900 gmsec-schema-level=1", Gmsec.DataType.KEY_VALUE_DATA) );
			Specification spec2 = new Specification( new Config("gmsec-specification-version=201900 gmsec-schema-level=2", Gmsec.DataType.KEY_VALUE_DATA) );

			check("Unexpected specification version", Gmsec.MSG_SPEC_2019_00 == spec0.getVersion());
			check("Unexpected specification schema level", Specification.SchemaLevel.LEVEL_0 == spec0.getSchemaLevel());

			check("Unexpected specification version", Gmsec.MSG_SPEC_2019_00 == spec1.getVersion());
			check("Unexpected specification schema level", Specification.SchemaLevel.LEVEL_1 == spec1.getSchemaLevel());

			check("Unexpected specification version", Gmsec.MSG_SPEC_2019_00 == spec2.getVersion());
			check("Unexpected specification schema level", Specification.SchemaLevel.LEVEL_2 == spec2.getSchemaLevel());
		}

		{
			Specification spec = new Specification( new Config("name=value", Gmsec.DataType.KEY_VALUE_DATA) );
			check("Unexpected specification version", spec.getVersion() == Gmsec.MSG_SPEC_CURRENT);
			check("Unexpected specification schema level", spec.getSchemaLevel() == Specification.SchemaLevel.LEVEL_0);
		}

		//o Off-nominal tests
		try {
			Specification spec = new Specification( new Config("gmsec-specification-version=201300", Gmsec.DataType.KEY_VALUE_DATA) );
			check("An exception was expected", false);
		}
		catch (GmsecException e) {
			check(e.getMessage(), e.getMessage().contains("Could not list files in template directory"));
		}

		try {
			Specification spec = new Specification( new Config("gmsec-specification-version=ABC", Gmsec.DataType.KEY_VALUE_DATA) );
			check("An exception was expected", false);
		}
		catch (GmsecException e) {
			check(e.getMessage(), e.getMessage().contains("\"GMSEC-SPECIFICATION-VERSION\" contains invalid value"));
		}

		try {
			Specification spec = new Specification( new Config("gmsec-schema-level=3", Gmsec.DataType.KEY_VALUE_DATA) );
			check("An exception was expected", false);
		}
		catch (GmsecException e) {
			check(e.getMessage(), e.getMessage().contains("DIRECTORY is missing definition for LEVEL-3 HEADER"));
		}

		try {
			Specification spec = new Specification( new Config("gmsec-schema-level=10", Gmsec.DataType.KEY_VALUE_DATA) );
			check("An exception was expected", false);
		}
		catch (GmsecException e) {
			check(e.getMessage(), e.getMessage().contains("\"GMSEC-SCHEMA-LEVEL\" contains out-of-range integer"));
		}

		try {
			Specification spec = new Specification( new Config("gmsec-schema-level=ABC", Gmsec.DataType.KEY_VALUE_DATA) );
			check("An exception was expected", false);
		}
		catch (GmsecException e) {
			check(e.getMessage(), e.getMessage().contains("\"GMSEC-SCHEMA-LEVEL\" contains invalid value"));
		}

		try {
			Specification spec = new Specification( new Config("gmsec-schema-path=here", Gmsec.DataType.KEY_VALUE_DATA) );
			check("An exception was expected", false);
		}
		catch (GmsecException e) {
			check(e.getMessage(), e.getMessage().contains("Could not list files in template directory"));
		}
	}


	private void test_copy_constructor()
		throws Exception
	{
		Log.info("Test copy-constructor");

		Specification spec1 = new Specification();
		Specification spec2 = new Specification(spec1);

		check("Unexpected specification version", spec1.getVersion() == spec2.getVersion());
		check("Unexpected specification schema level", spec1.getSchemaLevel() == spec2.getSchemaLevel());
		check("Unexpected message specifications", spec1.getMessageSpecifications().size() == spec2.getMessageSpecifications().size());
	}


	private void test_get_schema_id_iterator()
		throws Exception
	{
		Log.info("Test getSchemaIDIterator()");

		Specification spec = new Specification(); 

		SchemaIDIterator iter = spec.getSchemaIDIterator();

		check("Expected availability of schema IDs", iter.hasNext());
	}


	private void test_get_message_specifications()
		throws Exception
	{
		Log.info("Test getMessageSpecifications()");

		Specification spec = new Specification();

		Collection<MessageSpecification> msgSpecs = spec.getMessageSpecifications();

		check("Expected to have message specifications", msgSpecs.size() > 0);
	}


	private void test_get_header_field_names()
		throws Exception
	{
		Log.info("Test getHeaderFieldNames()");

		Specification spec = new Specification();

		Collection<String> headers = spec.getHeaderFieldNames();
		check("Expected to have message specifications", headers.size() == 26);

		Collection<String> headers2 = spec.getHeaderFieldNames("");
		check("Expected to have message specifications", headers2.size() == 26);

		Collection<String> headers3 = spec.getHeaderFieldNames(null);
		check("Expected to have message specifications", headers3.size() == 26);

		Collection<String> defaultHeaders = spec.getHeaderFieldNames("DEFAULT");
		check("Expected to have message specifications", headers.size() == 26);

		Collection<String> C2MSHeaders = spec.getHeaderFieldNames("C2MS");
		check("Expected to have message specifications", headers.size() == 26);
	}
}
